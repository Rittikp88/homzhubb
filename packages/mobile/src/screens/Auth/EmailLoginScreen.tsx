import React from 'react';
import { StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { bindActionCreators, Dispatch } from 'redux';
import { connect } from 'react-redux';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { UserSelector } from '@homzhub/common/src/modules/user/selectors';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { LoginForm } from '@homzhub/common/src/components/organisms/LoginForm';
import { Screen } from '@homzhub/mobile/src/components/HOC/Screen';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import {
  IEmailLoginPayload,
  ILoginFormData,
  ILoginPayload,
  LoginTypes,
} from '@homzhub/common/src/domain/repositories/interfaces';

interface IStateProps {
  isLoading: boolean;
}

interface IDispatchProps {
  login: (payload: ILoginPayload) => void;
}

type libraryProps = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.EmailLogin>;
type Props = IDispatchProps & IStateProps & libraryProps;

export class EmailLoginScreen extends React.PureComponent<Props> {
  public render(): React.ReactNode {
    const { t, isLoading } = this.props;
    return (
      <Screen
        containerStyle={styles.container}
        headerProps={{
          type: 'secondary',
          icon: icons.leftArrow,
          onIconPress: this.handleIconPress,
        }}
        pageHeaderProps={{
          contentTitle: t('auth:logInWithEmail'),
        }}
        backgroundColor={theme.colors.white}
        isLoading={isLoading}
      >
        <LoginForm
          isEmailLogin
          handleForgotPassword={this.handleForgotPassword}
          onLoginSuccess={this.handleLoginSuccess}
          testID="loginForm"
        />
      </Screen>
    );
  }

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private handleForgotPassword = (): void => {
    const {
      navigation,
      route: { params },
    } = this.props;
    const onCallback = params && params.onCallback ? { onCallback: params.onCallback } : {};
    navigation.navigate(ScreensKeys.ForgotPassword, onCallback);
  };

  private handleLoginSuccess = (values: ILoginFormData): void => {
    const {
      login,
      route: { params },
    } = this.props;

    const emailLoginData: IEmailLoginPayload = {
      action: LoginTypes.EMAIL,
      payload: {
        email: values.email,
        password: values.password,
      },
    };

    const loginPayload: ILoginPayload = {
      data: emailLoginData,
      ...(params && params.onCallback && { callback: params.onCallback }),
    };

    login(loginPayload);
  };
}

export const mapStateToProps = (state: IState): IStateProps => {
  return {
    isLoading: UserSelector.getLoadingState(state),
  };
};

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { login } = UserActions;
  return bindActionCreators(
    {
      login,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation()(EmailLoginScreen));

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
});
