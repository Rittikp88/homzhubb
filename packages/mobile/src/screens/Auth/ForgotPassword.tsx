import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { UserService } from '@homzhub/common/src/services/UserService';
import { theme } from '@homzhub/common/src/styles/theme';
import { FormButton, FormTextInput, AnimatedHeader, Text } from '@homzhub/common/src/components';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';

interface IForgotPasswordState {
  email: string;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.ForgotPassword>;

class ForgotPassword extends Component<Props, IForgotPasswordState> {
  public state = {
    email: '',
  };

  public render(): React.ReactNode {
    const { t } = this.props;
    const formData = { ...this.state };
    return (
      <View style={styles.container}>
        <AnimatedHeader
          icon="left-arrow"
          title={t('auth:forgotPassword')}
          subTitle={t('auth:forgotPasswordDescription')}
          onIconPress={this.handleIconPress}
          headerContainerStyle={styles.headerContainer}
        />
        <View style={styles.content}>
          <Formik onSubmit={this.onSubmit} validate={FormUtils.validate(this.formSchema)} initialValues={formData}>
            {(formProps: FormikProps<FormikValues>): React.ReactElement => (
              <>
                <FormTextInput
                  formProps={formProps}
                  inputType="email"
                  name="email"
                  label="Email"
                  placeholder={t('auth:enterEmail')}
                />
                <FormButton
                  formProps={formProps}
                  type="primary"
                  title={t('auth:forgotPasswordRequestLink')}
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  containerStyle={styles.formButtonStyle}
                />
              </>
            )}
          </Formik>
          <Text type="small" textType="semiBold" style={styles.backToLoginLink} onPress={this.navigateToLogin}>
            {t('auth:backToLogin')}
          </Text>
        </View>
      </View>
    );
  }

  private onSubmit = (formProps: IForgotPasswordState): void => {
    const { navigation } = this.props;
    const { email } = formProps;
    const payload = {
      action: 'SEND_EMAIL',
      payload: {
        email,
      },
    };
    UserService.resetPassword(payload)
      .then((response) => {
        const { email_exists, token } = response;
        if (email_exists) {
          navigation.navigate(ScreensKeys.ResetPassword, {
            token,
            email,
          });
        } else {
          AlertHelper.error({ message: 'Email does not exists' });
        }
        return null;
      })
      .catch((err) => {
        AlertHelper.error({ message: err });
      });
  };

  public navigateToLogin = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.EmailLogin);
  };

  public handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private formSchema = (): yup.ObjectSchema<{ email: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      email: yup.string().email(t('auth:emailValidation')).required(t('auth:emailRequired')),
    });
  };
}

export default withTranslation()(ForgotPassword);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  content: {
    paddingHorizontal: 24,
    paddingVertical: 16,
  },
  headerContainer: {
    borderBottomWidth: 0,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
  backToLoginLink: {
    color: theme.colors.blue,
    alignSelf: 'center',
  },
});
