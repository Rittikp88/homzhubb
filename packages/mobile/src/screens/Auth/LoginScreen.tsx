import React, { Component } from 'react';
import { withTranslation, WithTranslation } from 'react-i18next';
import { AnimatedHeader, LoginForm, SocialMediaComponent } from '@homzhub/mobile/src/components';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { ILoginFormData } from '@homzhub/common/src/domain/repositories/interfaces';

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.Login>;

export class LoginScreen extends Component<Props> {
  public render(): React.ReactNode {
    const { t, navigation } = this.props;

    return (
      <AnimatedHeader
        title={t('login')}
        subTitle={t('auth:newAroundHere')}
        linkText={t('auth:signup')}
        onIconPress={this.onClosePress}
        onLinkPress={this.onSignUpClicked}
        testID="headerEvents"
      >
        <>
          <LoginForm onLoginSuccess={this.onOtpLoginPress} testID="loginForm" />
          <SocialMediaComponent
            isFromLogin
            onEmailLogin={this.onEmailLoginPress}
            navigation={navigation}
            testID="socialEmailLogin"
          />
        </>
      </AnimatedHeader>
    );
  }

  private onClosePress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private onEmailLoginPress = (): void => {
    const {
      navigation,
      route: { params },
    } = this.props;
    const onCallback = params && params.onCallback ? { onCallback: params.onCallback } : {};
    navigation.navigate(ScreensKeys.EmailLogin, onCallback);
  };

  private onSignUpClicked = (): void => {
    const {
      navigation,
      route: { params },
    } = this.props;
    const onCallback = params && params.onCallback ? { onCallback: params.onCallback } : {};
    navigation.navigate(ScreensKeys.SignUp, onCallback);
  };

  private onOtpLoginPress = (values: ILoginFormData): void => {
    const {
      navigation,
      t,
      route: { params },
    } = this.props;
    navigation.navigate(ScreensKeys.OTP, {
      type: OtpNavTypes.Login,
      title: t('auth:loginOtp'),
      countryCode: values.phone_code,
      otpSentTo: values.phone_number,
      ...(params && params.onCallback && { onCallback: params.onCallback }),
    });
  };
}

export default withTranslation()(LoginScreen);
