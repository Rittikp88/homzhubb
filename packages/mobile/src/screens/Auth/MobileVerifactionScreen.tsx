import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { withTranslation, WithTranslation } from 'react-i18next';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { theme } from '@homzhub/common/src/styles/theme';
import { FormTextInput, AnimatedHeader, Text, FormButton } from '@homzhub/common/src/components';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { BottomSheetListView } from '@homzhub/mobile/src/components/molecules/BottomSheetListView';

interface IVerificationState {
  phone: string;
  countryCode: string;
  isBottomSheetVisible: boolean;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.MobileVerification>;

class MobileVerificationScreen extends Component<Props, IVerificationState> {
  public state = {
    phone: '',
    countryCode: '+91',
    isBottomSheetVisible: false,
  };

  public render(): React.ReactNode {
    const {
      t,
      route: {
        params: { icon, title, subTitle, buttonTitle, message },
      },
    } = this.props;
    const { isBottomSheetVisible, countryCode } = this.state;
    const formData = { ...this.state };
    return (
      <View style={styles.container}>
        <AnimatedHeader icon={icon} title={title} subTitle={subTitle} onIconPress={this.handleIconPress} />
        <View style={styles.content}>
          <Text type="small" style={styles.message}>
            {message}
          </Text>
          <Formik onSubmit={this.onSubmit} validate={FormUtils.validate(this.formSchema)} initialValues={formData}>
            {(formProps: FormikProps<FormikValues>): React.ReactElement => (
              <>
                <FormTextInput
                  formProps={formProps}
                  inputType="phone"
                  name="phone"
                  label="Phone"
                  inputPrefixText={countryCode}
                  onIconPress={this.handleDropdown}
                  placeholder={t('auth:yourNumber')}
                  helpText={t('auth:otpVerification')}
                />
                <FormButton
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  formProps={formProps}
                  type="primary"
                  title={buttonTitle}
                  containerStyle={styles.formButtonStyle}
                />
              </>
            )}
          </Formik>
        </View>
        <BottomSheetListView
          selectedValue={countryCode}
          listTitle={t('auth:countryRegion')}
          isBottomSheetVisible={isBottomSheetVisible}
          onCloseDropDown={this.onCloseDropDown}
          onSelectItem={this.handleSelection}
        />
      </View>
    );
  }

  private onSubmit = (): void => {
    // TOD0: Add logic
  };

  private onCloseDropDown = (): void => {
    this.setState({ isBottomSheetVisible: false });
  };

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private handleSelection = (value: string): void => {
    this.setState({ countryCode: value });
  };

  private handleDropdown = (): void => {
    const { isBottomSheetVisible } = this.state;
    this.setState({ isBottomSheetVisible: !isBottomSheetVisible });
  };

  private formSchema = (): yup.ObjectSchema<{ phone: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      phone: yup.string().required(t('auth:numberRequired')),
    });
  };
}

export default withTranslation()(MobileVerificationScreen);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  content: {
    paddingHorizontal: 24,
    paddingVertical: 16,
  },
  message: {
    color: theme.colors.darkTint3,
    marginBottom: 12,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
});
