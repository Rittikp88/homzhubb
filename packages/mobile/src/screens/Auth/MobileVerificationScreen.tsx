import React, { Component } from 'react';
import { StyleSheet, View } from 'react-native';
import { Formik, FormikHelpers, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { withTranslation, WithTranslation } from 'react-i18next';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { LocaleConstants } from '@homzhub/common/src/services/Localization/constants';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { SocialMediaKeys } from '@homzhub/common/src/assets/constants';
import { FormTextInput, FormButton, DetailedHeader, Text } from '@homzhub/common/src/components';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

interface IVerificationState {
  formData: {
    phone: string;
  };
  countryCode: string;
}

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.MobileVerification>;

export class MobileVerificationScreen extends Component<Props, IVerificationState> {
  public phone: FormTextInput | null = null;

  public state = {
    formData: {
      phone: '',
    },
    countryCode: '+91',
  };

  public render(): React.ReactNode {
    const { t } = this.props;
    const { formData, countryCode } = this.state;
    const { title, message, subTitle, buttonTitle } = this.getDisplayStrings();
    return (
      <View style={styles.container}>
        <DetailedHeader icon={icons.leftArrow} title={title} subTitle={subTitle} onIconPress={this.handleIconPress} />
        <View style={styles.content}>
          <Text type="small" style={styles.message}>
            {message}
          </Text>
          <Formik
            initialValues={{ ...formData }}
            onSubmit={this.onSubmit}
            validate={FormUtils.validate(this.formSchema)}
          >
            {(formProps: FormikProps<FormikValues>): React.ReactElement => (
              <>
                <FormTextInput
                  ref={(refs): void => {
                    this.phone = refs;
                  }}
                  formProps={formProps}
                  inputType="phone"
                  name="phone"
                  label="Phone"
                  inputPrefixText={countryCode}
                  onPhoneCodeChange={this.handlePhoneCodeChange}
                  phoneFieldDropdownText={t('auth:countryRegion')}
                  placeholder={t('yourNumber')}
                  helpText={t('otpVerification')}
                />
                <FormButton
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  formProps={formProps}
                  type="primary"
                  title={buttonTitle}
                  containerStyle={styles.formButtonStyle}
                />
              </>
            )}
          </Formik>
        </View>
      </View>
    );
  }

  private onSubmit = (values: { phone: string }, formActions: FormikHelpers<{ phone: string }>): void => {
    formActions.setSubmitting(true);
    const {
      t,
      route: {
        params: {
          isFromLogin,
          userData: {
            user: { first_name, last_name, email },
          },
          onCallback,
        },
      },
      navigation,
    } = this.props;
    const { countryCode } = this.state;
    const { phone } = values;

    navigation.navigate(ScreensKeys.OTP, {
      ref: () => this.phone,
      type: OtpNavTypes.SocialMedia,
      title: isFromLogin ? t('loginOtp') : t('verifyNumber'),
      phone,
      countryCode,
      userData: {
        full_name: `${first_name} ${last_name}`,
        email,
        phone_number: phone,
        phone_code: countryCode,
        // TODO (Aditya 10-Jun-2020): How to solve this password issue?
        password: 'RandomPassword',
      },
      ...(onCallback && { onCallback }),
    });
  };

  private handlePhoneCodeChange = (countryCode: string): void => {
    this.setState({ countryCode });
  };

  private handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private getDisplayStrings = (): { title: string; subTitle: string; message: string; buttonTitle: string } => {
    const {
      t,
      route: {
        params: { userData, isFromLogin },
      },
    } = this.props;
    const {
      provider,
      user: { first_name, email },
    } = userData;

    let title = t('signUpWithFacebook');
    const messageKey = isFromLogin ? 'enterNumberForProfileForLogin' : 'enterNumberForProfileForSignUp';

    if (isFromLogin) {
      if (provider === SocialMediaKeys.Google) {
        title = t('loginWithGoogle');
      } else {
        title = t('loginWithFacebook');
      }
    }

    if (!isFromLogin && provider === SocialMediaKeys.Google) {
      title = t('signUpWithGoogle');
    }

    return {
      title,
      subTitle: email,
      buttonTitle: isFromLogin ? t('common:login') : t('common:signUp'),
      message: t(`${messageKey}`, { givenName: first_name }),
    };
  };

  private formSchema = (): yup.ObjectSchema<{ phone: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      phone: yup.string().required(t('numberRequired')),
    });
  };
}

export default withTranslation(LocaleConstants.namespacesKey.auth)(MobileVerificationScreen);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  content: {
    paddingHorizontal: 24,
    paddingVertical: 16,
  },
  message: {
    color: theme.colors.darkTint3,
    marginBottom: 12,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
});
