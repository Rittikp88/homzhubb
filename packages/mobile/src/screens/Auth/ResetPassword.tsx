import React, { Component } from 'react';
import { View, StyleSheet } from 'react-native';
import { withTranslation, WithTranslation } from 'react-i18next';
import { Formik, FormikProps, FormikValues } from 'formik';
import * as yup from 'yup';
import { FormUtils } from '@homzhub/common/src/utils/FormUtils';
import { AlertHelper } from '@homzhub/mobile/src/utils/AlertHelper';
import { UserService } from '@homzhub/common/src/services/UserService';
import { theme } from '@homzhub/common/src/styles/theme';
import { icons } from '@homzhub/common/src/assets/icon';
import { FormButton, FormTextInput, DetailedHeader } from '@homzhub/common/src/components';
import { NavigationScreenProps, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';

type Props = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.ResetPassword>;

interface IResetPasswordState {
  password: string;
}

class ResetPassword extends Component<Props, IResetPasswordState> {
  public state = {
    password: '',
  };

  public render(): React.ReactNode {
    const { t } = this.props;
    const formData = { ...this.state };
    return (
      <View style={styles.container}>
        <DetailedHeader
          icon={icons.close}
          subTitle={t('auth:resetPassword')}
          subTitleType="large"
          subTitleColor={theme.colors.dark}
          onIconPress={this.handleIconPress}
          headerContainerStyle={styles.headerContainer}
        />
        <View style={styles.content}>
          <Formik onSubmit={this.onSubmit} validate={FormUtils.validate(this.formSchema)} initialValues={formData}>
            {(formProps: FormikProps<FormikValues>): React.ReactElement => (
              <>
                <FormTextInput
                  formProps={formProps}
                  inputType="password"
                  name="password"
                  label="Password"
                  helpText={t('auth:passwordValidation')}
                  placeholder={t('auth:enterPassword')}
                />
                <FormButton
                  formProps={formProps}
                  type="primary"
                  title={t('auth:resetPasswordButtonTitle')}
                  // @ts-ignore
                  onPress={formProps.handleSubmit}
                  containerStyle={styles.formButtonStyle}
                />
              </>
            )}
          </Formik>
        </View>
      </View>
    );
  }

  private onSubmit = (formProps: IResetPasswordState): void => {
    const { password } = formProps;
    const {
      navigation,
      route: {
        params: { token },
      },
    } = this.props;
    const payload = {
      action: 'SET_PASSWORD',
      payload: {
        token,
        password,
      },
    };
    UserService.resetPassword(payload)
      .then(() => {
        navigation.navigate(ScreensKeys.SuccessResetPassword);
        return null;
      })
      .catch((err) => {
        AlertHelper.error({ message: err });
      });
  };

  public handleIconPress = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.EmailLogin);
  };

  private formSchema = (): yup.ObjectSchema<{ password: string }> => {
    const { t } = this.props;
    return yup.object().shape({
      password: yup
        .string()
        .matches(FormUtils.passwordRegex, t('auth:passwordValidation'))
        .min(8, t('auth:minimumCharacters'))
        .required(t('auth:passwordRequired')),
    });
  };
}

export default withTranslation()(ResetPassword);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  content: {
    paddingHorizontal: 24,
  },
  headerContainer: {
    borderBottomWidth: 0,
  },
  formButtonStyle: {
    flex: 0,
    marginVertical: 30,
  },
});
