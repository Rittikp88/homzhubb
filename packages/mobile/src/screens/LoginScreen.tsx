import React, { Component } from 'react';
import { Animated, ScrollView, StyleSheet, View } from 'react-native';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { withTranslation, WithTranslation } from 'react-i18next';
import { IState } from '@homzhub/common/src/modules/interfaces';
import { UserActions } from '@homzhub/common/src/modules/user/actions';
import { IUserState } from '@homzhub/common/src/modules/user/interface';
import { ILoginFormData } from '@homzhub/common/src/domain/repositories/interfaces';
import { theme } from '@homzhub/common/src/styles/theme';
import { Header, LoginView } from '@homzhub/common/src/components';
import { AuthStackParamList } from '@homzhub/mobile/src/navigation/AuthStack';
import { NavigationScreenProps, OtpNavTypes, ScreensKeys } from '@homzhub/mobile/src/navigation/interfaces';

interface IDispatchProps {
  getSocialMedia: () => void;
}

interface IStateProps {
  user: IUserState;
}

interface ISignUpState {
  animatedValue: Animated.Value;
}

type libraryProps = WithTranslation & NavigationScreenProps<AuthStackParamList, ScreensKeys.Login>;
type Props = IStateProps & IDispatchProps & libraryProps;

class LoginScreen extends Component<Props, ISignUpState> {
  public state = {
    animatedValue: new Animated.Value(0),
  };

  public componentDidMount(): void {
    const { getSocialMedia } = this.props;
    getSocialMedia();
  }

  public render(): React.ReactNode {
    const { animatedValue } = this.state;
    const { t, user } = this.props;

    return (
      <View style={styles.container}>
        <ScrollView
          keyboardShouldPersistTaps="handled"
          scrollEventThrottle={16}
          showsVerticalScrollIndicator={false}
          onScroll={Animated.event([{ nativeEvent: { contentOffset: { y: animatedValue } } }], {
            useNativeDriver: false,
          })}
        >
          <View style={styles.scrollViewContent}>
            <LoginView
              onEmailLogin={this.handleEmailLoginPress}
              onSocialLoginSuccess={this.handleSocialLogin}
              socialMediaItems={user.data}
              handleLoginSuccess={this.handleLoginFormSubmit}
            />
          </View>
        </ScrollView>
        <Header
          isAnimation
          icon="close"
          title={t('login')}
          subTitle={t('auth:newAroundHere')}
          linkText={t('auth:signup')}
          animatedValue={animatedValue}
          onLinkPress={this.handleLinkPress}
          onIconPress={this.onClosePress}
        />
      </View>
    );
  }

  private onClosePress = (): void => {
    const { navigation } = this.props;
    navigation.goBack();
  };

  private handleEmailLoginPress = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.EmailLogin);
  };

  private handleLinkPress = (): void => {
    const { navigation } = this.props;
    navigation.navigate(ScreensKeys.SignUp);
  };

  private handleSocialLogin = (response: any): void => {
    const { navigation, t } = this.props;
    const {
      user: { email, givenName },
    } = response;
    navigation.navigate(ScreensKeys.MobileVerification, {
      title: t('auth:signUpWithGoogle') ?? '',
      subTitle: email,
      icon: 'left-arrow',
      message: t('auth:enterNumberForProfile', { givenName }) ?? '',
      buttonTitle: t('signUp'),
    });
  };

  private handleLoginFormSubmit = (values: ILoginFormData): void => {
    const { navigation, t } = this.props;
    navigation.navigate(ScreensKeys.OTP, {
      type: OtpNavTypes.Login,
      title: t('auth:loginOtp'),
      countryCode: values.country_code,
      phone: values.phone_number,
      ref: null,
    });
  };
}

const mapStateToProps = (state: IState): IStateProps => {
  const { user } = state;
  return {
    user,
  };
};

const mapDispatchToProps = (dispatch: Dispatch): IDispatchProps => {
  const { getSocialMedia } = UserActions;
  return bindActionCreators(
    {
      getSocialMedia,
    },
    dispatch
  );
};

export default connect<IStateProps, IDispatchProps, WithTranslation, IState>(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation()(LoginScreen));

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: theme.colors.white,
  },
  scrollViewContent: {
    marginTop: theme.headerConstants.headerMaxHeight,
  },
});
