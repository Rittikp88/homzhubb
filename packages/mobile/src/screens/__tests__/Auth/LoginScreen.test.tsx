import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { CommonRepository } from '@homzhub/common/src/domain/repositories/CommonRepository';
import { LoginScreen, mapDispatchToProps } from '@homzhub/mobile/src/screens/Auth/LoginScreen';
import { UserActionTypes } from '@homzhub/common/src/modules/user/actions';
import { SocialMediaData } from '@homzhub/common/src/mocks/socialMedia';

const mock = jest.fn();

describe('Login Screen', () => {
  let component: ShallowWrapper;
  let props: any;
  let instance: any;

  beforeEach(() => {
    props = {
      navigation: {
        navigate: mock,
        goBack: mock,
      },
      loginSuccess: mock,
    };
    component = shallow(
      <LoginScreen
        {...props}
        t={(key: string): string => key}
        route={{
          params: {
            onCallback: mock,
          },
        }}
      />
    );
    instance = component.instance();
  });

  it('should render login screen', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should navigate go back', () => {
    // @ts-ignore
    component.find('[testID="headerEvents"]').prop('onIconPress')();
    expect(mock).toHaveBeenCalled();
  });

  it('should navigate to signup screen', () => {
    // @ts-ignore
    component.find('[testID="headerEvents"]').prop('onLinkPress')();
    expect(mock).toHaveBeenCalled();
  });

  it('should navigate to email login', () => {
    // @ts-ignore
    component.find('[testID="socialEmailLogin"]').prop('onEmailLogin')();
    expect(mock).toHaveBeenCalled();
  });

  it('should login on otp press', () => {
    const values = {
      email: 'john@gmail.com',
      password: 'password',
      country_code: '_91',
      phone_number: '9876543210',
    };
    // @ts-ignore
    component.find('[testID="loginForm"]').prop('onLoginSuccess')(values);
    expect(mock).toHaveBeenCalled();
  });

  it('should call the mapDispatchToProps', () => {
    const dispatch = jest.fn();
    const data = {
      full_name: 'John doe',
      email: 'john@gmail.com',
      country_code: '_91',
      phone_number: '9876543210',
      access_token: 'access_token',
      refresh_token: 'refresh_token',
    };
    mapDispatchToProps(dispatch).loginSuccess(data);
    expect(dispatch.mock.calls[0][0]).toStrictEqual({
      type: UserActionTypes.AUTH.LOGIN_SUCCESS,
      payload: data,
    });
  });

  it('should fetch the social media data', async () => {
    jest.spyOn(CommonRepository, 'getSocialMedia').mockImplementation(() => Promise.resolve(SocialMediaData));
    await instance.componentDidMount();
    const response = await CommonRepository.getSocialMedia();
    component.setState({ socialMediaProviders: response });
    expect(component.state('socialMediaProviders')).toStrictEqual(SocialMediaData);
  });
});
