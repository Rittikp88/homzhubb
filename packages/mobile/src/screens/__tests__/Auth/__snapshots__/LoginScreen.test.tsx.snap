// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen should render login screen 1`] = `
<AnimatedHeader
  linkText="auth:signup"
  onIconPress={[Function]}
  onLinkPress={[Function]}
  subTitle="auth:newAroundHere"
  testID="headerEvents"
  title="login"
>
  <withI18nextTranslation(LoginForm)
    onLoginSuccess={[Function]}
    testID="loginForm"
  />
  <Connect(withI18nextTranslation(SocialMediaComponent))
    isFromLogin={true}
    navigation={
      Object {
        "goBack": [MockFunction],
        "navigate": [MockFunction],
      }
    }
    onEmailLogin={[Function]}
    testID="socialEmailLogin"
  />
</AnimatedHeader>
`;
